### PR TITLE
Add `setup` and `reset` lifecycle hooks + `replace` support

### PR DESCRIPTION
**Pull request by offirgolan**
- [x] Added `setup` lifecycle hook
- [x] Added `reset` lifecycle hook
- [x] Added support to declare `replace` from the QP map
- [x] Removed asserts for empty argument passed into `new QueryParams()`. This can allow anyone to take advantage of all the life cycle hooks even if they dont have QPs to declare.
- [x] Tests
- [x] Docs

With this PR we can actually eliminate the need for the route to declare any query params as well as handling data 🎉

### CODE REVIEW

**Comment by offirgolan** [comment]
@poteto @ebryn this is ready for review 😄

File: addon/instance-initializers/ember-parachute.js [hunk 1]
```diff
@@ -11,9 +11,64 @@ const {
   sendEvent
 } = Ember;
 
+const {
+  keys
+} = Object;
+
 export function initialize(/* application */) {
   Ember.Route.reopen({
```
Comment (ebryn) [inline addon/instance-initializers/ember-parachute.js]: We probably shouldn't be reopening in an instance initializer.

File: tests/integration/route-test.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+const EmberRoute = Ember.Route;
+const Controller = Ember.Controller.extend(queryParams.Mixin);
+let route;
+
+moduleFor('foo', 'Integration | Route', {
+  integration: true,
+
+  beforeEach() {
+    Ember.Route = Ember.Route.extend();
```
Comment (ebryn) [inline tests/integration/route-test.js]: bad

File: addon/instance-initializers/ember-parachute.js (hunk 1 above)
Comment (offirgolan) [inline addon/instance-initializers/ember-parachute.js]: @poteto what was the reason for switching over to an instance initializer?

**Comment by poteto** [comment]
As discussed let's do a readme overhaul in a separate PR

File: addon/initializers/ember-parachute.js [hunk 3]
```diff
@@ -11,9 +11,64 @@ const {
...
+      this._super(...arguments);
+
+      if (QueryParams.hasParachute(controller)) {
+        let { routeName } = this;
+        let event = new ParachuteEvent(routeName, controller, {});
+
+        // Overrides
+        event.shouldRefresh = false;
+
+        tryInvoke(controller, 'reset', [event, isExiting]);
```
Comment (poteto) [inline addon/initializers/ember-parachute.js]: I'm curious if we should even send a `ParachuteEvent` on `resetController`. Is there a scenario where the QPs change when the route resets? In the `reset` hook, the user can still access current QPs, reset QPs etc via the mixin. So it seems like we might be sending more than is necessary

Comment (offirgolan) [inline addon/initializers/ember-parachute.js]: I added it for the sake of consistency. That way you have access to all the same data in `setup`, `reset`, and `queryParamsDidChange`.

**Pull request merged**